clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java

```diff
@@ -0,0 +1,706 @@
+
+        // find the new consumer
+        String newConsumer = null;
+        for (String anotherConsumer: sortedCurrentSubscriptions) {
+            if (consumer2AllPotentialPartitions.get(anotherConsumer).contains(partition)) {
+                newConsumer = anotherConsumer;
+                break;
+            }
+        }
+
+        assert newConsumer != null;
+
+        // find the correct partition movement considering the stickiness requirement
+        TopicPartition partitionToBeMoved = partitionMovements.getTheActualPartitionToBeMoved(partition, consumer, newConsumer);
+        processPartitionMovement(partitionToBeMoved, newConsumer, sortedCurrentSubscriptions, currentPartitionConsumer);
+
+        return;
+    }
+
+    private void processPartitionMovement(TopicPartition partition, String newConsumer,
+                                          TreeSet<String> sortedCurrentSubscriptions,
+                                          HashMap<TopicPartition, String> currentPartitionConsumer) {
+        String oldConsumer = currentPartitionConsumer.get(partition);
+
+        sortedCurrentSubscriptions.remove(oldConsumer);
+        sortedCurrentSubscriptions.remove(newConsumer);
+
+        partitionMovements.movePartition(partition, oldConsumer, newConsumer);
+
+        currentAssignment.get(oldConsumer).remove(partition);
+        currentAssignment.get(newConsumer).add(partition);
+        currentPartitionConsumer.put(partition, newConsumer);
+        sortedCurrentSubscriptions.add(newConsumer);
+        sortedCurrentSubscriptions.add(oldConsumer);
+    }
+
+    @SuppressWarnings("serial")
+    private static class PartitionComparator implements Comparator<TopicPartition>, Serializable {
+        private Map<TopicPartition, List<String>> map;
+
+        PartitionComparator(Map<TopicPartition, List<String>> map) {
+            this.map = map;
+        }
+
+        @Override
+        public int compare(TopicPartition o1, TopicPartition o2) {
+            int ret = map.get(o1).size() - map.get(o2).size();
+            if (ret == 0) {
+                ret = o1.topic().compareTo(o2.topic());
+                if (ret == 0)
+                    ret = o1.partition() - o2.partition();
+            }
+            return ret;
+        }
+    }
+
+    @SuppressWarnings("serial")
+    protected static class SubscriptionComparator implements Comparator<String>, Serializable {
+        private Map<String, List<TopicPartition>> map;
+
+        SubscriptionComparator(Map<String, List<TopicPartition>> map) {
+            this.map = map;
+        }
+
+        @Override
+        public int compare(String o1, String o2) {
+            int ret = map.get(o1).size() - map.get(o2).size();
+            if (ret == 0)
+                ret = o1.compareTo(o2);
+            return ret;
+        }
+    }
+
+    @SuppressWarnings("serial")
+    protected static class PartitionAssignmentComparator implements Comparator<String>, Serializable {
```

Comment:
Just realized why I made these `Serializable`. It was because findbugs issues a warning:
>  SE_COMPARATOR_SHOULD_BE_SERIALIZABLE: Comparator doesn't implement Serializable
This class implements the Comparator interface. You should consider whether or not it should also implement the Serializable interface. If a comparator is used to construct an ordered collection such as a TreeMap, then the TreeMap will be serializable only if the comparator is also serializable. As most comparators have little or no state, making them serializable is generally easy and good defensive programming. 

What's the right approach here? Should we suppress them? Thanks.